lib/supavisor/handlers/proxy/db.ex

```diff
@@ -0,0 +1,236 @@
+              data.auth.secrets.(),
+              server_first_parts,
+              nonce,
+              data.auth.secrets.().user,
+              "biws"
+            )
+
+          bin = :pgo_protocol.encode_scram_response_message(client_final_message)
+          :ok = HH.sock_send(data.db_sock, bin)
+
+          {ps, :authentication_server_first_message, server_proof}
+
+        %{payload: {:authentication_server_first_message, server_first}}, {ps, _} ->
+          nonce = data.nonce
+          server_first_parts = :pgo_scram.parse_server_first(server_first, nonce)
+
+          {client_final_message, server_proof} =
+            :pgo_scram.get_client_final(
+              server_first_parts,
+              nonce,
+              data.auth.user,
+              data.auth.password.()
+            )
+
+          bin = :pgo_protocol.encode_scram_response_message(client_final_message)
+          :ok = HH.sock_send(data.db_sock, bin)
+
+          {ps, :authentication_server_first_message, server_proof}
+
+        %{payload: {:authentication_server_final_message, _server_final}}, acc ->
+          acc
+
+        %{payload: {:authentication_md5_password, salt}}, {ps, _} ->
+          Logger.debug("ProxyDb: dec_pkt, #{inspect(dec_pkt, pretty: true)}")
+
+          digest =
+            if data.auth.method == :password do
+              H.md5([data.auth.password.(), data.auth.user])
+            else
+              data.auth.secrets.().secret
+            end
+
+          payload = ["md5", H.md5([digest, salt]), 0]
+          bin = [?p, <<IO.iodata_length(payload) + 4::signed-32>>, payload]
+          :ok = HH.sock_send(data.db_sock, bin)
+          {ps, :authentication_md5}
+
+        %{tag: :error_response, payload: error}, _ ->
+          {:error_response, error}
+
+        _e, acc ->
+          acc
+      end)
+
+    case resp do
+      {_, :authentication_sasl, nonce} ->
+        {:keep_state, %{data | nonce: nonce}}
+
+      {_, :authentication_server_first_message, server_proof} ->
+        {:keep_state, %{data | server_proof: server_proof}}
+
+      {_, :authentication_md5} ->
+        {:keep_state, data}
+
+      {:error_response, ["SFATAL", "VFATAL", "C28P01", reason, _, _, _]} ->
+        Logger.error("ProxyDb: Auth error #{inspect(reason)}")
+        {:stop, :invalid_password, data}
+
+      {:error_response, error} ->
+        Logger.error("ProxyDb: Error auth response #{inspect(error)}")
+        {:keep_state, data}
+
+      {:ready_for_query, ps, db_state} ->
+        msg = "ProxyDb: DB ready_for_query: #{inspect(db_state)} #{inspect(ps, pretty: true)}"
+        Logger.debug(msg)
+        ps_encoded = Server.encode_parameter_status(ps)
+
+        {:next_state, :idle, %{data | parameter_status: ps},
+         {:next_event, :internal, {:client, {:greetings, ps_encoded}}}}
+
+      other ->
+        Logger.error("ProxyDb: Undefined auth response #{inspect(other)}")
+        {:stop, :auth_error, data}
+    end
+  end
```

Comment:
That is way to long and way to complex function to be a single one.

Reply:
can you please clarify?

Reply:
That function has 140 lines of code.

Reply:
ahh, gotcha